pkg/topology/scheduler/autoplacetopology/autoplacetopology.go

```diff
@@ -0,0 +1,148 @@
+//
+// This scheduler works like autoplace.Scheduler with a few key differences:
+// * If a `Requisite` topology is requested, all placement will be restricted onto those nodes.
+// * If a `Preferred` topology is requested, this scheduler will try to place at least one volume on those nodes.
+// This scheduler complies with the CSI spec for topology:
+//   https://github.com/container-storage-interface/spec/blob/v1.4.0/csi.proto#L523
+type Scheduler struct {
+	*lc.HighLevelClient
+	log *logrus.Entry
+}
+
+// Ensure Scheduler conforms to scheduler.Interface.
+var _ scheduler.Interface = &Scheduler{}
+
+func (s *Scheduler) Create(ctx context.Context, vol *volume.Info, req *csi.CreateVolumeRequest) error {
+	log := s.log.WithField("volume", vol.ID)
+
+	params, err := volume.NewParameters(vol.Parameters)
+	if err != nil {
+		return fmt.Errorf("unable to determine volume parameters: %w", err)
+	}
+
+	// The CSI spec mandates:
+	// * If `Requisite` exists, we _have_ to use those up first.
+	// * If `Requisite` and `Preferred` exists, we have `Preferred` ⊆ `Requisite`, and `Preferred` SHOULD be used first.
+	// * If `Requisite` does not exist and `Preferred` exists, we SHOULD use `Preferred`.
+	// * If both `Requisite` and `Preferred` do not exist, we can do what ever.
+	//
+	// Making this compatible with LINSTOR autoplace parameters can be quite tricky. For example, a "DoNotPlaceWith"
+	// constraint could mean we are not able to place the volume on the first preferred node.
+	//
+	// This scheduler works by first trying to place a volume on one of the preferred nodes. This should optimize
+	// placement in case of late volume binding (where the first preferred node is the node starting the pod).
+	// Then, normal autoplace happens, restricted to the requisite nodes. If there are still replicas to schedule, use
+	// autoplace again, this time without extra placement constraints from topology.
+
+	topos := req.GetAccessibilityRequirements()
+	log.WithField("requirements", topos).Trace("got topology requirement")
+
+	for _, preferred := range topos.GetPreferred() {
+		node, ok := preferred.GetSegments()[topology.LinstorNodeKey]
+		if !ok {
+			log.WithField("segment", preferred.GetSegments()).Trace("segment without node name, skipping")
+			continue
+		}
+
+		log.WithField("preferred", node).Trace("try initial placement on a preferred node")
+
+		// Just try a single autoplace request on one of the preferred nodes if possible. We use AutoPlace instead
+		// of MakeAvailable to ensure we respect the user-defined constraint from the storage class.
+		err := s.Resources.Autoplace(ctx, vol.ID, lapi.AutoPlaceRequest{
+			SelectFilter: lapi.AutoSelectFilter{PlaceCount: 1, NodeNameList: []string{node}},
+		})
+		if err != nil {
+			log.WithError(err).WithField("preferred", node).Trace("failed to autoplace")
+			continue
+		}
+
+		log.WithField("preferred", node).Trace("successfully placed on preferred node")
+
```

Comment:
Nitpicking but it puzzled me a bit when seeing that code and finding that it "alwas" `break`s in the happy path. maybe put it in an `else` (to the `err != nil`), or just add a comment like `// successfully autoplaced, we are done`, or maybe just remove the empty line so that it is close to the `log.WithField(... successfully)`. Something like this